### PR TITLE
[tune] Make save/restore developer APIs

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Trainable.save/restore have been used as developer APIs in practice - there is no reason for users to directly interact with these. This PR marks them as developer APIs, promotes the `_TrainingResult` (the return value) to developer API, and changes the save/restore API to always emit a `TrainingResult` and accept them in return.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/__init__.py [hunk 1]
```diff
@@ -39,5 +45,6 @@
     "RunConfig",
     "ScalingConfig",
     "TrainingIterator",
+    "TrainingResult",
```
Comment (justinvyu) [inline python/ray/train/__init__.py:48]: Does this need to be aliased at the top level? Can we just keep it developer API in an internal module?

File: python/ray/train/_internal/session.py [hunk 2]
```diff
@@ -74,13 +74,14 @@ class TrialInfo:
 
 
 @dataclass
-class TrainingResult:
+class _InternalTrainingResult:
```
Comment (justinvyu) [inline python/ray/train/_internal/session.py:77]: this one's to be deleted soon. Maybe `_LegacyTrainingResult` so we can find all the `Legacy` things

File: python/ray/tune/execution/tune_controller.py [hunk 3]
```diff
@@ -1988,10 +1994,10 @@ def _process_trial_save(
         """
         logger.debug("Trial %s: Processing trial save.", trial)
 
-        from ray.train._internal.checkpoint_manager import _TrainingResult
+        from ray.train._internal.checkpoint_manager import TrainingResult
 
         try:
-            if _use_storage_context() and isinstance(checkpoint_value, _TrainingResult):
+            if _use_storage_context() and isinstance(checkpoint_value, TrainingResult):
```
Comment (justinvyu) [inline python/ray/tune/execution/tune_controller.py:2000]: change isinstance check to be an assertion inside the if block

File: python/ray/tune/execution/tune_controller.py [hunk 4]
```diff
@@ -2014,6 +2020,9 @@ def _process_trial_save(
                 self._checkpoint_manager.on_trial_checkpoint(trial)
                 self._mark_trial_to_checkpoint(trial)
             else:
+                if isinstance(checkpoint_value, TrainingResult):
+                    checkpoint_value = checkpoint_value.checkpoint.path
+
```
Comment (justinvyu) [inline python/ray/tune/execution/tune_controller.py:2025]: then can remove this in the old codepath.

File: python/ray/tune/trainable/trainable.py [hunk 5]
```diff
@@ -935,25 +939,17 @@ def restore(
             )
             return True
 
+        if isinstance(checkpoint, TrainingResult):
+            checkpoint = checkpoint.checkpoint
+
```
Comment (justinvyu) [inline python/ray/tune/trainable/trainable.py:944]: no need to modify stuff below

File: python/ray/tune/trainable/trainable.py [hunk 6]
```diff
@@ -1030,15 +1026,22 @@ def restore_from_object(self, obj):
             "Use Trainable.restore() instead."
         )
 
-    def delete_checkpoint(self, checkpoint_path: Union[str, Checkpoint]):
+    def delete_checkpoint(
+        self, checkpoint_path: Union[str, LegacyCheckpoint, NewCheckpoint]
+    ):
```
Comment (justinvyu) [inline python/ray/tune/trainable/trainable.py:1031]: Let's hard deprecate this API. Was only needed to schedule remote deletions, but that happens via the storage filesystem directly now.

**Review by justinvyu — COMMENTED** [review]
Thanks. Comments are mostly about old codepath again, and we can remove the `Trainable.delete_checkpoint` API in this PR as well.

File: python/ray/train/__init__.py (hunk 1 above)
Comment (krfricke) [reply python/ray/train/__init__.py:48]: I think DeveloperAPI should be in a public module. I agree, we should probably keep the top-level `train` namespace clean.

Maybe `train.trainer.TrainingResult`, but I'll just keep it in the internal module for now and we can see later.

File: python/ray/tune/trainable/trainable.py (hunk 5 above)
Comment (krfricke) [reply python/ray/tune/trainable/trainable.py:944]: Mostly deleting the node syncing

**Comment by krfricke** [comment]
Let's defer this after 2.7 (and update after we picked all remaining fixes)

**Pull request closed without merging**